Add basic multi-panel capabilities

marohrdanz opened the pull request:
The main change in this PR is adding basic capabilities to initialize the panel layout in the NGCHM Viewer from this R package.

To create an .ngchm file with panels initialized, the ShaidyMapGen.jar changes from https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/553 are also required.

This PR includes only the first of the requested items in issue #17.  

Basic usage of this new feature is documented in the constructor for `panel_configuration` in the new file `panelClasses.R`.

In addition, this PR includes a few other smaller changes:

1. Moved logging initialization to `initNGCHM` (e286cc3, files functions.R, zzz.R)
2. Removed unused arguments from `writeProperties` (67a0234, this is the only change in file allMethods.R)
3. Add extra space in print.shaidyRepo message (f78e62d, this is the only change in file shaidy.R)
4. Allow error message arg to `castListToInteger()` (617cde9, file function.R)
5. Bump version number: 1.0.4 -> 1.1.0 (be3abe5, in DESCRIPTION)

## Testing performed

I verified that:

- A map built with this code and ShaidyMapGen.jar from https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/553 opens correctly in the current NGCHM Viewer version
- A map built using this R code but NOT the corresponding ShaidyMapGen.jar changes builds OK and opens in the NGCHM Viewer (just w/o the panels initialized)

See also the Testing Performed section in ShaidyMapGen.jar PR: https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/553

## Demo Script For Reviewers

To facilitate review, here is a sample R script to use this new feature:

```r
##
## Modify as needed to run on your system.
##

## Feature requires Java changes in https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/553
Sys.setenv("SHAIDYMAPGEN" = <path to ShaidyMapGen.jar built from NGCHM Viewer PR 553>)

nrow <- 30
ncol <- 30
matrix_data <- matrix(rnorm(nrow * ncol), nrow = nrow, ncol = ncol)
rownames(matrix_data) <- paste("Row", 1:nrow)
colnames(matrix_data) <- paste("Col", 1:ncol)


# Create the two pieces for the panels: 1. `panes_list` 2. `pane_types`
#
# 1. This `panes_list` corresponds to a three panel layout, where the horizontal
#    space is split 50/50, pane2 occupies 30% of the vertical space of the right
#    half, and pane3 occupies 70% of the vertical space of the right half:
# 
#             | pane2 
#       pane1 | -----
#             | pane3
#
panes_list <- list(pane(id="pane1"),
                   list(pane(id="pane2"),
                        pane(id="pane3"),
                        sizes = c(30, 70)),
                   sizes = c(50, 50))

# 2. `pane_types` specifies the type for each pane
#
#  WARNING: there is no error checking for the 'pluginName'. 
#           These are plugins defined in the NGCHM Viewer's custom.js file, or, 
#           in the case of special coordinates, generated automatically
#           from the name of the special coordinate. In this example, the column
#           covariate 'PCA.coordinate.*' gives rise to an automatically generated
#           plugin with name '2D ScatterPlot: PCA (column)'.  
#          (Feature added in https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/546)
#
pane_types <- list(
  pane1 = detailMap(id = "pane1"),
  pane2 = summaryMap(id = "pane2"),
  pane3 = pluginPane(id = "pane3", pluginName = "2D ScatterPlot: PCA (column)")
)

# These two pieces are arguments to `panel_configuration`:
panel_config <- panel_configuration(panes_list, pane_types)

# Create chm with specified `panel_configuration`:
hm <- chmNew(myname, matrix_data, panel_configuration = panel_config)

# Add PCA covariates
pca_data <- prcomp(as.data.frame(t(matrix_data)), scale = TRUE, center = TRUE, rank = 10)
hm <- chmAddPCA(hm, "column", pca_data, basename = "PCA")

# Add discrete covariate (e.g. to be used for coloring a scatter plot)
covariate_vector <- sample(c("ONE", "TWO"), ncol, replace = TRUE)
names(covariate_vector) <- colnames(matrix_data)
covariateBar <- chmNewCovariate("Group", covariate_vector)
hm <- chmAddCovariateBar(hm, "col", covariateBar)

# Write .ngchm file that will open in viewer w/ 3 panels
chmExportToFile(hm, "panels_from_R_demo.ngchm", overwrite = TRUE)
```

marohrdanz committed:
Move log initialization to initNGCHM

This is a more sensible place to initialize logging. Probably logging
will only be used by developers, so maybe we don't need/want to
initialize it anywhere--developers can initialize it themselves if
desired.

Also imported trace-level logging.

marohrdanz committed:
Remove unused arguments from writeProperties

Arguments 'saveDir' and 'chm@format' were not used in the function.
Removed them for clarity.

marohrdanz committed:
Implement default panel layout

This implements a basic version of the feature request to initialize a
panel layout from R. This does not include all of the requested items in
issue #17. Indeed, the only new layout supported is a three-panel layout
with geometry:

          | pane2
    pane1 | ------
          | pane3

where the user can specify the content of each pane (e.g. pane1 is a
summary map, pane2 is a plugin, etc). The current defaul layout is also
supported:

   pane1 (detailMap) | pane2 (summaryMap)

NOTE: This functionality also depends on ShaidyMapGen.jar. The changes
in https://github.com/MD-Anderson-Bioinformatics/NG-CHM/pull/553 are
also required in order to produce a .ngchm file with panels initialized.

NOTE: This introduces unfortunate additional coupling to the NGCHM
viewer project. This is because of the lines like:

```javascript
  mapItem.dataBoxWidth = savedState.dataBoxWidth;
  ...
  let zoomBoxSizeIdx = DET.zoomBoxSizes.indexOf(mapItem.dataBoxWidth);
```

in DET.restoreFromSavedState (see DetailHeatMapDisplay.js). Because of
this, `dataBoxWidth` must both be defined and one of the
DET.zoomBoxSizes. I'm not sure there's a good way around this, given
that we don't want to introduce breaking changes into the NGCHM
ecosystem.

marohrdanz committed:
Bump version number: 1.0.4 -> 1.1.0

We have added a backwards-compatible feature: very basic multi-panel
layout options.

marohrdanz commented:
I know we are still determining exactly what initial features and code quality we want for an initial implementation. Marked as 'ready for review' with anticipation of discussion and requests for changes.

ChrisWakefield reviewed (approved):
I'm ok with this, having reviewed most source changes minus the .Rd.

jmelott reviewed (approved):
Noting that this is just an initial version of this feature, and does not allow users to set up any configuration of panels that they might desire.

marohrdanz commented on R/functions.R:
```diff
@@ -4413,3 +4411,140 @@ verifyNumeric <- function(variableToCheck) {
+    vertical = FALSE,
+    children = list(pane1, container2)
+  )
+  panel_layout <- panel_layout(
```
From @bmbroom : `panel_layout <- list("pane1", list("pane2", "pane3"))`

With the goal of this being interpreted as a top-level horizontal container, with a pane1 on the left, and a vertical container on the right, with pane2 above pane3. (same as example in code).

marohrdanz committed:
Add initial version of panel layout API

This allows the user to specify panel layout. Basic usage is documented
in the constructor for `panel_configuration`.

This is part of issue #17.

marohrdanz committed:
Allow error message arg to castListToInteger()

I want to use this function, but the default R error message probably
won't help my users. So adding the option to provide a custom error
message.

marohrdanz committed:
Add extra space in print.shaidyRepo message

Just a tweak to improve readability of the message.

marohrdanz commented:
Merging after additional approval/discussion in today's meeting.
